frmts/avif/avifdataset.cpp

```diff
@@ -0,0 +1,1178 @@
+
+/* static */ avifResult GDALAVIFIO::Read(struct avifIO *io, uint32_t readFlags,
+                                         uint64_t offset, size_t size,
+                                         avifROData *out)
+{
+    GDALAVIFIO *gdalIO = reinterpret_cast<GDALAVIFIO *>(io);
+    if (readFlags != 0)
+    {
+        // Unsupported readFlags
+        return AVIF_RESULT_IO_ERROR;
+    }
+    if (offset > gdalIO->nFileSize)
+    {
+        return AVIF_RESULT_IO_ERROR;
+    }
+
+    if (offset == gdalIO->nFileSize)
+    {
+        out->data = gdalIO->buffer.data();
+        out->size = 0;
+        return AVIF_RESULT_OK;
+    }
+
+    const uint64_t availableSize = gdalIO->nFileSize - offset;
+    size = static_cast<size_t>(std::min<uint64_t>(size, availableSize));
+    try
+    {
+        gdalIO->buffer.resize(size);
+    }
+    catch (const std::bad_alloc &)
+    {
+        CPLError(CE_Failure, CPLE_OutOfMemory,
+                 "Out of memory in GDALAVIFIO::Read()");
+        return AVIF_RESULT_IO_ERROR;
+    }
+
+    if (gdalIO->fp->Seek(offset, SEEK_SET) != 0 ||
+        gdalIO->fp->Read(gdalIO->buffer.data(), size, 1) != 1)
+    {
+        return AVIF_RESULT_IO_ERROR;
+    }
+
+    out->data = gdalIO->buffer.data();
+    out->size = size;
+    return AVIF_RESULT_OK;
+}
+
+/************************************************************************/
+/*                              Init()                                  */
+/************************************************************************/
+
+bool GDALAVIFDataset::Init(GDALOpenInfo *poOpenInfo)
+{
+    m_decoder = avifDecoderCreate();
+    if (!m_decoder)
+        return false;
+
+    std::string osFilename(poOpenInfo->pszFilename);
+    VSIVirtualHandleUniquePtr fp(poOpenInfo->fpL);
+    poOpenInfo->fpL = nullptr;
+
+    if (STARTS_WITH_CI(poOpenInfo->pszFilename, "AVIF:"))
+    {
+        const char *pszPartPos = poOpenInfo->pszFilename + strlen("AVIF:");
+        const char *pszNextColumn = strchr(pszPartPos, ':');
+        if (pszNextColumn == nullptr)
+            return false;
+        m_iPart = atoi(pszPartPos);
+        if (m_iPart <= 0)
+            return false;
+        osFilename = pszNextColumn + 1;
+        fp.reset(VSIFOpenL(osFilename.c_str(), "rb"));
+        if (fp == nullptr)
+            return false;
+    }
+
+    auto gdalIO = std::make_unique<GDALAVIFIO>(std::move(fp));
+    avifDecoderSetIO(m_decoder, reinterpret_cast<avifIO *>(gdalIO.release()));
+
+    const auto avifErr = avifDecoderParse(m_decoder);
+    if (avifErr != AVIF_RESULT_OK)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "avifDecoderParse() failed with: %s",
+                 avifResultToString(avifErr));
+        return false;
+    }
+
+    // AVIF limit is 65,536 x 65,536 pixels;
+    nRasterXSize = static_cast<int>(m_decoder->image->width);
+    nRasterYSize = static_cast<int>(m_decoder->image->height);
+
+    if (m_decoder->image->depth > 16)
```

Comment:
```suggestion
    if (m_decoder->image->depth > 12)
```
?

Reply:
Both are fine, as the GDAL part  would be ready for [12-16] bits, but changed to 12 to avoid any surprise